sql/rowexec: TestUncertaintyErrorIsReturned failed

[(sql/rowexec).TestUncertaintyErrorIsReturned failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2766465&tab=buildLog) on [master@fc68d927f946103a757888fcbd3ac20904c57a6b](https://github.com/cockroachdb/cockroach/commits/fc68d927f946103a757888fcbd3ac20904c57a6b):

```
      <autogenerated>:1 +0x57
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:382 +0x2e4
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run()
      <autogenerated>:1 +0x84
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:367 +0x7a7
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:980 +0x290
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1036 +0x99b
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:908 +0xc19
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:652 +0x14cd
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:120 +0xf53
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:207 +0x1d3
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1583 +0x732
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1585 +0x4c5
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1437 +0x565
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482 +0x139
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:625 +0x3f0

Goroutine 79 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:553 +0x1fd
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:285 +0x832
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:152 +0x304
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:658 +0xeb7
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2074 +0x16c
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:156 +0x11d
==================
    testing.go:1038: race detected during execution of test
        --- FAIL: TestUncertaintyErrorIsReturned/vectorize=true/SELECT_*_FROM_t_AS_t1_JOIN_t_AS_t2_ON_t1.x_=_t2.x (0.08s)
=== RUN   TestUncertaintyErrorIsReturned/vectorize=true
=== CONT  TestUncertaintyErrorIsReturned/vectorize=true
    testing.go:1038: race detected during execution of test
    --- FAIL: TestUncertaintyErrorIsReturned/vectorize=true (0.21s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestUncertaintyErrorIsReturned PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestUncertaintyErrorIsReturned.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/util/timeutil/stopwatch.go
+++ b/pkg/util/timeutil/stopwatch.go
@@ -11,24 +11,28 @@
 package timeutil
 
 import (
-	"sync/atomic"
 	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 )
 
 // StopWatch is a utility stop watch that can be safely started and stopped
 // multiple times and can be used concurrently.
 type StopWatch struct {
-	// started is used as boolean where 0 indicates false and 1 indicates true.
-	// started is "true" if the stop watch has been started.
-	started int32
-	// startedAt is the time when the stop watch was started.
-	startedAt time.Time
-	// elapsed is the total time measured by the stop watch (i.e. between all
-	// Starts and Stops).
-	elapsed time.Duration
-	// timeSource is the source of time used by the stop watch. It is always
-	// timeutil.Now except for tests.
-	timeSource func() time.Time
+	mu struct {
+		syncutil.Mutex
+		// started is true if the stop watch has been started and haven't been
+		// stopped after that.
+		started bool
+		// startedAt is the time when the stop watch was started.
+		startedAt time.Time
+		// elapsed is the total time measured by the stop watch (i.e. between
+		// all Starts and Stops).
+		elapsed time.Duration
+		// timeSource is the source of time used by the stop watch. It is always
+		// timeutil.Now except for tests.
+		timeSource func() time.Time
+	}
 }
 
 // NewStopWatch creates a new StopWatch.
@@ -43,13 +47,18 @@ func NewTestStopWatch(timeSource func() time.Time) *StopWatch {
 }
 
 func newStopWatch(timeSource func() time.Time) *StopWatch {
-	return &StopWatch{timeSource: timeSource}
+	w := &StopWatch{}
+	w.mu.timeSource = timeSource
+	return w
 }
 
 // Start starts the stop watch if it hasn't already been started.
 func (w *StopWatch) Start() {
-	if atomic.CompareAndSwapInt32(&w.started, 0, 1) {
-		w.startedAt = w.timeSource()
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if !w.mu.started {
+		w.mu.started = true
+		w.mu.startedAt = w.mu.timeSource()
 	}
 }
 
@@ -57,14 +66,19 @@ func (w *StopWatch) Start() {
 // the duration that elapsed since it was started. If the stop watch has
 // already been stopped, it is a noop.
 func (w *StopWatch) Stop() {
-	if atomic.CompareAndSwapInt32(&w.started, 1, 0) {
-		w.elapsed += w.timeSource().Sub(w.startedAt)
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.mu.started {
+		w.mu.started = false
+		w.mu.elapsed += w.mu.timeSource().Sub(w.mu.startedAt)
 	}
 }
 
 // Elapsed returns the total time measured by the stop watch so far.
 func (w *StopWatch) Elapsed() time.Duration {
-	return w.elapsed
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.mu.elapsed
 }
 
 // TestTimeSource is a source of time that remembers when it was created (in

